### PR TITLE
Publish Rresource API integration

### PR DESCRIPTION
**Pull request by priyanka-TL**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: generics/services/scp.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * author : Priyanka Pradeep
+ * Date : 20-dec-2024
+ * Description : All scp related api call.
+ */
+
+//dependencies
+const request = require('request')
+
+// Function to callback to SCP service
+const resourcePublishCallBack = function (callBackUrl, resourceId, templateId) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/scp.js:12]: @priyanka-TL please add function signature

File: generics/services/scp.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	return new Promise(async (resolve, reject) => {
+		try {
+			// Construct the URL for the user service
+			const url = `${callBackUrl}?resource_id=${resourceId}&published_id=${templateId.toString()}`
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/scp.js:16]: @priyanka-TL what will be the url? can you share one sample pls

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Reviewed on Dec 24 12:35

File: module/project/templates/helper.js [hunk 3]
```diff
@@ -1526,3 +1632,281 @@ function _taskAndSubTaskinSequence(query, projectionValue, language = '') {
...
+	}
+}
+
+/**
+ * Create and Find Categories
+ * @name _getOrCreateCategories
+ * @param {Object} categories - Categories Data
+ * @returns {Object} - Response contains categories data
+ */
+function _getOrCreateCategories(categories) {
```
[the remark this answers is not in the document]
Comment (priyanka-TL) [reply module/project/templates/helper.js:1754]: We have that capability in SCP

File: module/project/templates/helper.js [hunk 4]
```diff
@@ -1526,3 +1632,281 @@ function _taskAndSubTaskinSequence(query, projectionValue, language = '') {
...
+/**
+ * Create Task
+ * @name createTasks
+ * @param {Object} tasks - task data
+ * @param {String} templateId - template Id
+ * @param {String} templateExternalId - template externaldId
+ * @param {String} parentId - parentId
+ * @returns {Object} - Response contains task data
+ */
+async function createTasks(tasks, templateId, templateExternalId, parentId = null) {
```
[the remark this answers is not in the document]
Comment (priyanka-TL) [reply module/project/templates/helper.js:1844]: yes

File: constants/interface-routes/configs.json [hunk 5]
```diff
@@ -1702,6 +1702,19 @@
 					"packageName": "elevate-project"
 				}
 			]
+		},
+		{
+			"sourceRoute": "/project/v1/project/templates/publishTemplateAndTasks",
```
Comment (aks30) [inline constants/interface-routes/configs.json:1707]: @priyanka-TL Let's create a new module SCP and then put all relevant APIs there.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 24 Dec

File: constants/interface-routes/configs.json (hunk 5 above)
Comment (priyanka-TL) [reply constants/interface-routes/configs.json]: @aks30 done

File: module/scp/helper.js [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+const projectTemplateTaskQueries = require(DB_QUERY_BASE_PATH + '/projectTemplateTask')
+const projectCategoriesQueries = require(DB_QUERY_BASE_PATH + '/projectCategories')
+const scpService = require(GENERICS_FILES_PATH + '/services/scp')
+
+/**
+ * SCP Helper
+ * @class
+ */
+
+module.exports = class ReportsHelper {
```
Comment (VISHNUDAS-tunerlabs) [inline module/scp/helper.js:19]: @priyanka-TL Is this helper name relatable?

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]
Reviewed On Dec 26

**Pull request merged**